rust/gui-client/src-tauri/src/client/gui/system_tray_menu.rs

```diff
@@ -3,146 +3,164 @@
 //! "Notification Area" is Microsoft's official name instead of "System tray":
 //! <https://learn.microsoft.com/en-us/windows/win32/shell/notification-area?redirectedfrom=MSDN#notifications-and-the-notification-area>
 
-use connlib_client_shared::callbacks::ResourceDescription;
-use std::str::FromStr;
+use connlib_client_shared::callbacks::{ResourceDescription, Status};
+use serde::{Deserialize, Serialize};
 use tauri::{CustomMenuItem, SystemTrayMenu, SystemTrayMenuItem, SystemTraySubmenu};
+use url::Url;
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
 pub(crate) enum Event {
+    AdminPortal,
     CancelSignIn,
-    Resource { id: String },
+    Copy(String),
     SignIn,
     SignOut,
     ShowWindow(Window),
+    Url(Url),
     Quit,
 }
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
 pub(crate) enum Window {
     About,
     Settings,
 }
 
-#[derive(thiserror::Error, Debug)]
-pub(crate) enum Error {
-    #[error("the system tray menu item ID is not valid")]
-    InvalidId,
-}
-
-const ABOUT_KEY: &str = "/about";
-const SETTINGS_KEY: &str = "/settings";
-const QUIT_KEY: &str = "/quit";
-const QUIT_ACCELERATOR: &str = "Ctrl+Q";
-
-impl FromStr for Event {
-    type Err = Error;
-
-    fn from_str(s: &str) -> Result<Self, Error> {
-        Ok(match s {
-            ABOUT_KEY => Self::ShowWindow(Window::About),
-            "/cancel_sign_in" => Self::CancelSignIn,
-            SETTINGS_KEY => Self::ShowWindow(Window::Settings),
-            "/sign_in" => Self::SignIn,
-            "/sign_out" => Self::SignOut,
-            QUIT_KEY => Self::Quit,
-            s => {
-                let id = s.strip_prefix("/resource/").ok_or(Error::InvalidId)?;
-                Self::Resource { id: id.to_string() }
-            }
-        })
-    }
-}
+const QUIT_TEXT_SIGNED_OUT: &str = "Quit Firezone";
 
 pub(crate) fn signed_in(user_name: &str, resources: &[ResourceDescription]) -> SystemTrayMenu {
     let mut menu = SystemTrayMenu::new()
-        .add_item(
-            CustomMenuItem::new("".to_string(), format!("Signed in as {user_name}")).disabled(),
-        )
-        .add_item(CustomMenuItem::new("/sign_out".to_string(), "Sign out"))
-        .add_native_item(SystemTrayMenuItem::Separator)
-        .add_item(CustomMenuItem::new("".to_string(), "Resources").disabled());
+        .disabled(format!("Signed in as {user_name}"))
+        .item(Event::SignOut, "Sign out")
+        .separator()
+        .disabled("Resources");
 
     for res in resources {
-        let id = res.id();
-        let submenu = SystemTrayMenu::new().add_item(CustomMenuItem::new(
-            format!("/resource/{id}"),
-            res.pastable(),
-        ));
+        let submenu = SystemTrayMenu::new();
+
+        let address_description = res.address_description();
```

Comment:
What happens if the address_description is blank or not set? It's an optional field. In the other clients we show the address instead.